relay-server/src/actors/project_cache.rs

```diff
@@ -519,30 +615,90 @@ impl ProjectCacheBroker {
         &mut self,
         message: CheckEnvelope,
     ) -> Result<CheckedEnvelope, DiscardReason> {
-        let project = self.get_or_create_project(message.envelope.meta().public_key());
-
+        let CheckEnvelope { envelope, context } = message;
+        let project = self.get_or_create_project(envelope.meta().public_key());
         // Preload the project cache so that it arrives a little earlier in processing. However,
         // do not pass `no_cache`. In case the project is rate limited, we do not want to force
         // a full reload. Fetching must not block the store request.
         project.prefetch(false);
+        project.check_envelope(envelope, context)
+    }
+
+    /// Handles the processing of the provided envelope.
+    fn handle_processing(
+        &mut self,
+        project_key: ProjectKey,
+        state: Arc<ProjectState>,
+        sampling_state: Option<Arc<ProjectState>>,
+        envelope: Box<Envelope>,
+        envelope_context: EnvelopeContext,
+    ) {
+        // The `Envelope` and `EnvelopeContext` will be dropped if the `Project::check_envelope()`
+        // function returns any error, which will also be ignored here.
+        let Some(Ok(checked)) = self
+            .projects
+            .get_mut(&project_key)
+            .map(|p| p.check_envelope(envelope, envelope_context)) else { return; };
+
+        if let Some((envelope, envelope_context)) = checked.envelope {
+            let mut process = ProcessEnvelope {
+                envelope,
+                envelope_context,
+                project_state: state.clone(),
+                sampling_project_state: None,
+            };
 
-        project.check_envelope(message.envelope, message.context)
+            if let Some(sampling_state) = sampling_state {
+                if state.organization_id == sampling_state.organization_id {
+                    process.sampling_project_state = Some(sampling_state)
+                }
+            }
+
+            EnvelopeProcessor::from_registry().send(process);
+        }
     }
 
+    /// Checks an incoming envelope and decides either process it immediately or buffer it.
+    ///
+    /// Few conditions are checked here:
+    /// - If there is no dynamic sampling key and the project is already cached, we do strait to
+    /// processing otherwise buffer the envelopes.
+    /// - If there is a dynamic sampling key is provided, if the root and sampling project
+    /// are cached - process the envelope, buffer otherwise.
```

Comment:
```suggestion
    /// - If there a dynamic sampling key is provided, and the root and sampling project
    /// are cached - process the envelope, buffer otherwise.
```